File: 03a.py
#!/usr/bin/python3

with open('03.dat') as f:
    data = f.read()
numbers = data.split('\n')
numbers.pop()

ones = [0]*len(numbers[0])
zeros = [0]*len(numbers[0])
gamma = ""
epsilon = ""
for number in numbers:
    for index, digit in enumerate(number):
        if digit == '0':
            try:
                zeros[index] += 1
            except IndexError:
                zeros[index] = 1
        else:
            try:
                ones[index] += 1
            except IndexError:
                ones[index] = 1
for index, _ in enumerate(zeros):
    if (zeros[index] > ones[index]):
        gamma += "0"
        epsilon += "1"
    else:
        gamma += "1"
        epsilon += "0"
print(int(gamma, 2)*int(epsilon,2))

File: 03a.js
import * as fs from 'fs';

const filename = "03.dat";
fs.readFile(filename, (err, data) => {
    if (err) throw err;

    var number_len = /(\d+)/.exec(data.toString()).at(0).length;
    var zeros = [];
    var ones = [];
    for (var i = 0; i < number_len; i++) {
        zeros.push(0);
        ones.push(0);
    }

    var gamma = "";
    var epsilon = "";

    data.toString().split("\n").forEach(number => {
        for (var i = 0; i < number.length; i++) {
            if (number[i] === '0') {
                zeros[i]++;
            } else {
                ones[i]++;
            }
        }
    });

    for (var i = 0; i < number_len; i++) {
        if (zeros[i] > ones[i]) {
            gamma += "0";
            epsilon += "1";
        }
        else {
            gamma += "1";
            epsilon += "0";
        }
    }
    console.log(parseInt(gamma, 2)*parseInt(epsilon,2));
});
